File: src/PythonClient/lib/quixstreaming/models/streamreader/streampropertiesreader.py
from typing import Dict

from quixstreaming.eventhook import EventHook
from datetime import datetime, timedelta
from quixstreaming.helpers import *

from quixstreaming import __importnet
import Quix.Sdk.Streaming


class StreamPropertiesReader(object):

    def __init__(self, net_object: Quix.Sdk.Streaming.Models.StreamReader.StreamPropertiesReader):
        """
            Initializes a new instance of StreamPropertiesReader.
            NOTE: Do not initialize this class manually, use StreamReader.properties to access an instance of it

            Parameters:

            net_object (.net object): The .net object representing a StreamPropertiesReader
        """
        if net_object is None:
            raise Exception("StreamPropertiesReader is none")
        else:
            self.__wrapped = net_object

        self.on_changed = EventHook(name="StreamPropertiesReader.on_changed")
        """
        Raised when stream properties changed. This is an event without parameters. Interrogate the class raising
        it to get new state
        """

        def __on_changed_net_handler():
            self.on_changed.fire()
        self.__wrapped.OnChanged += __on_changed_net_handler

    @property
    def name(self) -> str:
        """ Gets the name of the stream """
        return self.__wrapped.Name

    @property
    def location(self) -> str:
        """ Gets the location of the stream """
        return self.__wrapped.Location

    @property
    def time_of_recording(self) -> datetime:
        """ Gets the datetime of the recording """
        return NetToPythonConverter.convert_datetime(self.__wrapped.TimeOfRecording)

    @property
    def metadata(self) -> Dict[str, str]:
        """ Gets the metadata of the stream """
        return NetToPythonConverter.convert_dict(self.__wrapped.Metadata)

    @property
    def parents(self) -> datetime:
        """ Gets the metadata of the stream """
        return NetToPythonConverter.convert_list(self.__wrapped.Parents)

File: src/CsharpClient/Quix.Sdk.Streaming/Models/StreamReader/StreamPropertiesReader.cs
using System;
using System.Collections.Generic;
using Quix.Sdk.Process.Models;

namespace Quix.Sdk.Streaming.Models.StreamReader
{
    /// <summary>
    /// Properties and Metadata of the stream.
    /// All the changes of these properties are populated to this class automatically
    /// </summary>
    public class StreamPropertiesReader : IDisposable
    {
        private readonly IInputTopic topic;
        private readonly IStreamReaderInternal streamReader;

        /// <summary>
        /// Initializes a new instance of <see cref="StreamPropertiesReader"/>
        /// </summary>
        /// <param name="topic">The topic the stream to what this reader belongs to</param>
        /// <param name="streamReader">Stream reader owner</param>
        internal StreamPropertiesReader(IInputTopic topic, IStreamReaderInternal streamReader)
        {
            this.topic = topic;
            this.streamReader = streamReader;

            this.Metadata = new Dictionary<string, string>();
            this.Parents = new List<string>();

            this.streamReader.OnStreamPropertiesChanged += OnStreamPropertiesChangedEventHandler;

        }

        private void OnStreamPropertiesChangedEventHandler(IStreamReader sender, StreamProperties streamProperties)
        {
            this.Name = streamProperties.Name;
            this.Location = streamProperties.Location;
            this.TimeOfRecording = streamProperties.TimeOfRecording;
            this.Metadata = streamProperties.Metadata;
            this.Parents = streamProperties.Parents;

            this.OnChanged?.Invoke(this.streamReader, new StreamPropertiesChangedEventArgs(this.topic, sender));
        }

        /// <summary>
        /// Raised when the stream properties have changed
        /// Sender is the stream the properties changed for
        /// </summary>
        public event EventHandler<StreamPropertiesChangedEventArgs> OnChanged;

        /// <summary>
        /// Gets the name of the stream
        /// </summary>
        public string Name { get; private set; }

        /// <summary>
        /// Gets the location of the stream
        /// </summary>
        public string Location { get; private set; }

        /// <summary>
        /// Gets the datetime of the recording
        /// </summary>
        public DateTime? TimeOfRecording { get; private set; }

        /// <summary>
        /// Gets the metadata of the stream
        /// </summary>
        public Dictionary<string, string> Metadata { get; private set; }

        /// <summary>
        /// List of Stream Ids of the Parent streams
        /// </summary>
        public List<string> Parents { get; private set; }

        /// <inheritdoc/>
        public void Dispose()
        {
            this.streamReader.OnStreamPropertiesChanged -= OnStreamPropertiesChangedEventHandler;
        }
    }

    public class StreamPropertiesChangedEventArgs
    {
        public StreamPropertiesChangedEventArgs(IInputTopic topic, IStreamReader reader)
        {
            this.Topic = topic;
            this.Stream = reader;
        }

        public IInputTopic Topic { get; }
        public IStreamReader Stream { get; }
    }
}
